nn/nnclient.py

```diff
@@ -0,0 +1,305 @@
+       but WITHOUT ANY WARRANTY; without even the implied warranty of
+       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+       GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/.
+
+
+    This module implements a client that connects to a middleware
+    neural network server (see nnserver/nnserver.py), which in turn
+    connects to a TensorFlow model server.
+
+"""
+
+import json
+import requests
+import logging
+
+from settings import Settings
+import tokenizer
+
+from nn import nntree
+from nn.nntree import ParseResult
+from reynir.binparser import BIN_Token
+from reynir import bintokenizer
+
+
+def index_text(text):
+    """ Segments contiguous (Icelandic) text into paragraphs and sentences
+        and returns:
+            dictionary of sentence indices to sentences
+            dictionary of paragraph index to constituent sentence indices"""
+    text = "[[ " + " ]] [[ ".join(text.split("\n")) + " ]]"
+    tok_stream = bintokenizer.tokenize(text)
+
+    pgs = tokenizer.paragraphs(tok_stream)
+    pg_idx_to_sent_idx = dict()
+    sent_idx_to_sent = dict()
+    curr_sent_idx = 0
+    curr_pg_idx = 0
+
+    for pg in pgs:
+        sent_idxs = []
+        for (idx, sent) in pg:
+            curr_sent = list(filter(BIN_Token.is_understood, sent))
+            curr_sent = " ".join([tok.txt for tok in curr_sent])
+            sent_idxs.append(curr_sent_idx)
+            sent_idx_to_sent[curr_sent_idx] = curr_sent
+            curr_sent_idx += 1
+        pg_idx_to_sent_idx[curr_pg_idx] = sent_idxs
+        curr_pg_idx += 1
+    return pg_idx_to_sent_idx, sent_idx_to_sent
+
+
+class NnClient:
+    """ A client that connects to the HTTP rest interface of
+        a tensorflow model server (using plaintext) """
+
+    port = None
+    host = None
+    verb = None
+
+    @classmethod
+    def request_sentence(cls, text):
+        """ Request neural network output for a single sentence """
+        raise NotImplemented
+
+    @classmethod
+    def request_text(cls, text):
+        """ Request neural network output for contiguous text """
+        raise NotImplemented
+
+    @classmethod
+    def _request(cls, pgs, data=None):
+        """ Send serialized request to remote model server """
+        url = "http://{host}:{port}/{verb}.api".format(
+            host=cls.host, port=cls.port, verb=cls.verb
+        )
+        headers = {"content-type": "application/json"}
+        payload = {"pgs": pgs}
+
+        if data is not None:
+            for (k, v) in data.items():
+                payload[k] = v
+
+        logging.debug(str(payload))
+        payload = json.dumps(payload)
+        resp = requests.post(url, data=payload, headers=headers)
+        resp.raise_for_status()
+
+        obj = json.loads(resp.text)
+        if "valid" in obj:
```

Comment:
Is this definitely correct as an error condition?

Reply:
@haukurb ?